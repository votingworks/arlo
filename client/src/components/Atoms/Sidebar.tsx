import React from 'react'
import styled from 'styled-components'
import { Menu, Spinner, IMenuItemProps } from '@blueprintjs/core'
import H2Title from './H2Title'

const Wrapper = styled.div`
  margin-right: 30px;
  width: 250px;

  ul.bp3-menu {
    padding-left: 0;
  }
`

export interface ISidebarMenuItem {
  id: string
  activate?: (e?: unknown | null, force?: boolean) => void
  title: string
  active: boolean
  state: 'live' | 'processing' | 'locked'
}

interface IProps {
  menuItems: ISidebarMenuItem[]
  title: string
}

const Sidebar = ({ menuItems, title }: IProps) => (
  <Wrapper>
    <H2Title>{title}</H2Title>
    <Menu>
      {menuItems.map((item, i) => {
        const itemProps: IMenuItemProps = {
          active: item.active,
          text: item.title,
          disabled: item.state !== 'live',
          labelElement:
            item.state === 'processing' ? (
              <Spinner size={Spinner.SIZE_SMALL} />
            ) : null,
        }
        if (item.activate) itemProps.onClick = item.activate
        return (
          <React.Fragment key={item.id}>
            {i > 0 && <Menu.Divider />}
            <Menu.Item {...itemProps} />
          </React.Fragment>
        )
      })}
    </Menu>
  </Wrapper>
)

export default Sidebar
