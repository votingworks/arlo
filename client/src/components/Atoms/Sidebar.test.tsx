import React from 'react'
import { render } from '@testing-library/react'
import Sidebar, { ISidebarMenuItem } from './Sidebar'

const mockMenuItems: ISidebarMenuItem[] = [
  {
    title: 'Item One',
    activate: jest.fn(),
    active: false,
    state: 'live',
  },
  {
    title: 'Item Two',
    activate: jest.fn(),
    active: false,
    state: 'locked',
  },
  {
    title: 'Item Three',
    activate: jest.fn(),
    active: false,
    state: 'processing',
  },
  {
    title: 'Item Four',
    activate: jest.fn(),
    active: true,
    state: 'live',
  },
]

describe('Sidebar', () => {
  it('renders all options', () => {
    const { container } = render(
      <Sidebar menuItems={mockMenuItems} title="Test Sidebar" />
    )
    expect(container).toMatchSnapshot()
  })
})
