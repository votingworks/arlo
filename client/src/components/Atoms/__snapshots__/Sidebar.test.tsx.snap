// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders all options 1`] = `
<div>
  <div
    class="sc-bwzfXH bzeSVq"
  >
    <h2
      class="bp3-heading sc-bdVaJa ibVpPq"
    >
      Test Sidebar
    </h2>
    <ul
      class="bp3-menu"
    >
      <li
        class=""
      >
        <a
          class="bp3-menu-item bp3-popover-dismiss"
        >
          <div
            class="bp3-text-overflow-ellipsis bp3-fill"
          >
            Item One
          </div>
        </a>
      </li>
      <li
        class="bp3-menu-divider"
      />
      <li
        class=""
      >
        <a
          class="bp3-menu-item bp3-disabled"
          tabindex="-1"
        >
          <div
            class="bp3-text-overflow-ellipsis bp3-fill"
          >
            Item Two
          </div>
        </a>
      </li>
      <li
        class="bp3-menu-divider"
      />
      <li
        class=""
      >
        <a
          class="bp3-menu-item bp3-disabled"
          tabindex="-1"
        >
          <div
            class="bp3-text-overflow-ellipsis bp3-fill"
          >
            Item Three
          </div>
          <span
            class="bp3-menu-item-label"
          >
            <div
              class="bp3-spinner"
            >
              <div
                class="bp3-spinner-animation"
              >
                <svg
                  height="20"
                  stroke-width="16.00"
                  viewBox="-3.00 -3.00 106.00 106.00"
                  width="20"
                >
                  <path
                    class="bp3-spinner-track"
                    d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                  />
                  <path
                    class="bp3-spinner-head"
                    d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                    pathLength="280"
                    stroke-dasharray="280 280"
                    stroke-dashoffset="210"
                  />
                </svg>
              </div>
            </div>
          </span>
        </a>
      </li>
      <li
        class="bp3-menu-divider"
      />
      <li
        class=""
      >
        <a
          class="bp3-menu-item bp3-active bp3-intent-primary bp3-popover-dismiss"
        >
          <div
            class="bp3-text-overflow-ellipsis bp3-fill"
          >
            Item Four
          </div>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
